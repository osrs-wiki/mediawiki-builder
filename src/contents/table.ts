import MediaWikiContent from "../content";

export type MediaWikiTableRow = {
  cells: MediaWikiTableCell[];
  header?: boolean;
  options?: MediaWikiCellOptions;
};

export type MediaWikiTableCell = {
  content: MediaWikiContent[];
  options?: MediaWikiCellOptions;
};

export type MediaWikiTableParams = {
  caption?: string;
  rows: MediaWikiTableRow[];
  options?: MediaWikiTableOptions;
};

export type MediaWikiTableOptions = {
  class?: string;
  style?: string;
};

export type MediaWikiCellOptions = {
  class?: string;
  colspan?: string;
  rowspan?: string;
  style?: string;
};

export class MediaWikiTable extends MediaWikiContent {
  caption?: string;
  options?: MediaWikiTableOptions;
  rows: MediaWikiTableRow[];

  constructor(params: MediaWikiTableParams) {
    super();
    this.caption = params.caption;
    this.rows = params.rows;
    this.options = params.options;
  }

  build() {
    const tableOptions = this.options
      ? Object.keys(this.options).reduce(
          (total, key) =>
            `${total} ${key}="${
              this.options?.[key as keyof MediaWikiTableOptions]
            }"`,
          ""
        )
      : "";
    return `{|${tableOptions}\n${
      this.caption ? `|+${this.caption}\n` : ""
    }${this.rows
      .map(
        (row) =>
          `|-\n` +
          row.cells
            .map(
              (cell) =>
                (row.header ? `!` : `|`) +
                cell.content.map((content) => content.build()).join("") +
                "\n"
            )
            .join("")
      )
      .join("")}|}`;
  }
}
