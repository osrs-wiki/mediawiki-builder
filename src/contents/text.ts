import MediaWikiContent from "../content";

export type MediaWikiTextStyling = {
  italics?: boolean;
  bold?: boolean;
  underline?: boolean;
};

export class MediaWikiText extends MediaWikiContent {
  value: string;
  styling?: MediaWikiTextStyling;

  constructor(value: string, styling?: MediaWikiTextStyling) {
    super();
    this.value = value;
    this.styling = styling;
  }

  build() {
    if (this.value === "") {
      return this.value;
    }
    if (this.styling?.bold) {
      this.value = `'''${this.value}'''`;
    }
    if (this.styling?.italics) {
      this.value = `''${this.value}''`;
    }
    if (this.styling?.underline) {
      this.value = `<u>${this.value}</u>`;
    }
    return `${this.value}`;
  }
}
