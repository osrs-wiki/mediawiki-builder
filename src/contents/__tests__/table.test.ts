import { MediaWikiTable } from "../table";
import { MediaWikiText } from "../text";

describe("MediaWikiTable", () => {
  test("it should build correctly with no options", () => {
    const table = new MediaWikiTable({
      rows: [
        {
          header: true,
          cells: [
            {
              content: [new MediaWikiText("test1")],
            },
            {
              content: [new MediaWikiText("test2")],
            },
          ],
        },
        {
          cells: [
            {
              content: [
                new MediaWikiText("test1", { bold: true }),
                new MediaWikiText(" table1"),
              ],
            },
            {
              content: [
                new MediaWikiText("test2", { bold: true }),
                new MediaWikiText(" table2"),
              ],
            },
          ],
        },
        {
          cells: [
            {
              content: [
                new MediaWikiText("test3", { italics: true }),
                new MediaWikiText(" table3"),
              ],
            },
            {
              content: [
                new MediaWikiText("test4\n"),
                new MediaWikiText(" table4"),
              ],
            },
          ],
        },
      ],
    });
    expect(table.build()).toMatchSnapshot();
  });

  test("it should build correctly with table options", () => {
    const table = new MediaWikiTable({
      rows: [
        {
          header: true,
          cells: [
            {
              content: [new MediaWikiText("header")],
            },
          ],
        },
        {
          cells: [
            {
              content: [new MediaWikiText("value")],
            },
          ],
        },
      ],
      options: {
        class: "wikitable",
      },
    });
    expect(table.build()).toMatchSnapshot();
  });
});
