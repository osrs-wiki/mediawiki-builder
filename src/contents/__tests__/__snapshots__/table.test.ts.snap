// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaWikiTable it should build correctly with no options 1`] = `
"{|
|-
!test1
!test2
|-
|'''test1''' table1
|'''test2''' table2
|-
|''test3'' table3
|test4
 table4
|}"
`;

exports[`MediaWikiTable it should build correctly with table options 1`] = `
"{| class=\\"wikitable\\"
|-
!header
|-
|value
|}"
`;
